src/domain/communication/message-inbox/message.inbox.service.ts

```diff
@@ -0,0 +1,339 @@
+      this.logger.verbose?.(
+        `No other members in DIRECT room ${room.id}, skipping`,
+        LogContext.COMMUNICATION
+      );
+      return;
+    }
+
+    // Find all VCs among other members
+    const vcMembers = [];
+    for (const actorID of otherMembers) {
+      const vc =
+        await this.virtualContributorLookupService.getVirtualContributorByAgentId(
+          actorID
+        );
+      if (vc) {
+        vcMembers.push(actorID);
+      }
+    }
+
+    if (vcMembers.length === 0) {
+      this.logger.verbose?.(
+        `No VC members found in DIRECT room ${room.id}, skipping`,
+        LogContext.COMMUNICATION
+      );
+      return;
+    }
+
+    // Build AgentInfo from sender
+    const agentInfo = await this.buildAgentInfo(payload.actorID);
+
+    // Use message ID as threadID for DIRECT conversations
+    const threadID = payload.message.id;
+
+    // Invoke all VCs in parallel
+    this.logger.verbose?.(
+      `Invoking ${vcMembers.length} VC(s) in DIRECT conversation`,
+      LogContext.COMMUNICATION
+    );
+
+    await Promise.all(
+      vcMembers.map(vcActorID =>
+        this.virtualContributorMessageService.invokeVirtualContributor(
+          vcActorID,
+          payload.message.message,
+          threadID,
+          agentInfo,
+          '', // contextSpaceID out of scope
+          room,
+          {
+            id: generateVcInteractionId(threadID, vcActorID),
+            threadID,
+            virtualContributorID: vcActorID,
+          }
+        )
+      )
+    );
+  }
+
+  private async handleExistingThread(
+    payload: any,
+    room: any,
+    threadID: string,
+    vcData: any
+  ): Promise<void> {
+    // Check if sender is NOT the main VC (avoid infinite loop)
+    if (payload.actorID === vcData.virtualContributorActorID) {
+      this.logger.verbose?.(
+        `Ignoring message from VC itself (actorID=${payload.actorID})`,
+        LogContext.COMMUNICATION
+      );
+      return;
+    }
+
+    // Invoke the main VC for this thread
+    this.logger.verbose?.(
+      `Invoking VC for existing thread: threadID=${threadID}, vcActorID=${vcData.virtualContributorActorID}`,
+      LogContext.COMMUNICATION
+    );
+
+    // Build AgentInfo from sender
+    const agentInfo = await this.buildAgentInfo(payload.actorID);
+
+    await this.virtualContributorMessageService.invokeVirtualContributor(
+      vcData.virtualContributorActorID,
+      payload.message.message,
+      threadID,
+      agentInfo,
+      '', // contextSpaceID out of scope
+      room,
+      {
+        id: generateVcInteractionId(threadID, vcData.virtualContributorActorID),
+        threadID,
+        virtualContributorID: vcData.virtualContributorActorID,
+      }
+    );
+  }
+
+  /**
+   * Build AgentInfo from actorID (agent.id).
+   * Attempts to find user by agent.id, falls back to anonymous if not found (e.g., VC sender).
+   */
+  private async buildAgentInfo(actorID: string): Promise<AgentInfo> {
```

Comment:
This must be part of the AgentInfo service, not deep down into a piece of the business logic.